### PR TITLE
Update scripts

### PR DESCRIPTION
**Pull request by glebm**
Managed to deploy this today :) https://176.58.116.34/instructions
This is an attempt to extract some deployment and production provisioning configuration, not ideal but a start. 
- [x] docker:
  -  had to [add `rootflags=nobarrier`](https://github.com/fgrehm/vagrant-lxc/wiki/Usage-on-Linode#configuring-grub)  to GRUB get Docker working with AUFS on Linode (using pv-grub).
- [x] default http in production
- [x] production / deployment env in bash 
- [x] remove .ruby-gemset
- [x] remove redundant echo statements

### CODE REVIEW

File: .ruby-gemset [hunk 1]
```diff
@@ -0,0 +1 @@
+contestrus
```
Comment (sirupsen) [inline .ruby-gemset]: I'd rather not have this in here. People should vendor their gems with bundler or run in the VM where this doesn't really matter.

File: script/deploy [hunk 2]
```diff
@@ -14,6 +16,7 @@ if [[ ! "$REPLY" =~ ^[Yy]$ ]]; then
   exit 1
 fi
 
+echo "As $DEPLOY_USER on $DEPLOY_HOST:"
```
Comment (sirupsen) [inline script/deploy]: Not really necessary imo. when we have `set -x` set.

File: script/deploy [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
-DEPLOY_USER="app"
-DEPLOY_HOST="contestrus"
+# deploy with git copy
+
+source $(dirname "$0")/deploy-env
 
 DEPLOY_TARGET="${1:-HEAD}"
 SHA="$(git rev-parse "$DEPLOY_TARGET")"
 ABBREV_SHA="${SHA:0:10}"
+REPO_URL=${REPO_URL:-"`git config --get remote.origin.url`"}
```
Comment (sirupsen) [inline script/deploy:10]: Is this so forks work as expected?

File: script/deploy-env [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+cd "$(dirname "$0")/.."
+CONFIG_FILE='config/deploy.yml'
+[ -f $CONFIG_FILE ] && eval $(ruby -ryaml -e "puts *YAML.load_file('${CONFIG_FILE}').to_a.map {|x| x * '='}")
```
Comment (sirupsen) [inline script/deploy-env]: I think we should get rid of those YAML files and just make it bash. It's so much simpler than having to do things like this and the syntax doesn't change very much.

**Comment by sirupsen** [comment]
Thanks a lot for your contribution @glebm!

I've left a few comments. @charliesome and @bouk you guys care taking a look too?

File: .ruby-gemset (hunk 1 above)
Comment (haileys) [inline .ruby-gemset]: Yeah this app already isolates its own gems, no need for gemsets

File: script/provision-basic [hunk 5]
```diff
@@ -35,7 +35,7 @@ apt-get install -y \
   sudo \
   vim \
   zlib1g-dev \
-  linux-image-extra-`uname -r`
+  linux-image-extra-3.11.0-13-generic
```
Comment (sirupsen) [inline script/provision-basic]: why was this changed?

**Comment by sirupsen** [comment]
Your deployment is quite fresh in memory and this adds a few extra steps. Would you mind contributing to the deployment section in the README?

File: script/provision-basic (hunk 5 above)
Comment (glebm) [inline script/provision-basic]: `uname -r` is not very reliable: I was deploying this on linode, where `uname -r` for the default Ubuntu 13.10 distro returns `3.11.6-x86_64-linode35`.

File: script/deploy-env (hunk 4 above)
Comment (glebm) [inline script/deploy-env]: Agreed, this is ugly. In Bash, would we need to `export` every variable or is there another way? Not really that familiar with bash best practices

File: script/deploy (hunk 3 above)
Comment (glebm) [inline script/deploy:10]: Yeah, before it was hardcoded to Sirupsen/contestrus

File: .ruby-gemset (hunk 1 above)
Comment (glebm) [inline .ruby-gemset]: I believe that a gemset is always a good thing. I added it originally when it installed everything into my global gemset when running bundle the first time (before vagrant), and a gemset would avoid that. Will remove if you don't think it's a valid case.

Comment (sirupsen) [inline .ruby-gemset]: You can't really develop locally unless you're using a recent Linux kernel for development. Everyone should use Vagrant to develop Contestrus. Setting it up locally is not trivial because of Docker.

Comment (bouk) [inline .ruby-gemset]: Nah, development should all be done inside Vagrant because then docker and such is not correctly set up

**Comment by glebm** [comment]
In my deployment docker seems to take up all the available disk space; it uses vfs, not sure if that's OK.

File: script/deploy-env (hunk 4 above)
Comment (sirupsen) [inline script/deploy-env]: Narh, you don't need to export them all if you just source this in the start of deploy.

File: script/deploy (hunk 3 above)
Comment (sirupsen) [inline script/deploy:10]: Cool, I'm good with this

File: config/deploy.yml.sample [hunk 6]
```diff
@@ -0,0 +1,2 @@
+DEPLOY_USER: app
+DEPLOY_HOST: contestrus
```
Comment (bouk) [inline config/deploy.yml.sample]: Please add a trailing newline

File: script/provision-basic (hunk 5 above)
Comment (sirupsen) [inline script/provision-basic]: Hm.. I'm worried about forgetting this when we upgrade to 14.04..

File: config/deploy.yml.sample (hunk 6 above)
Comment (glebm) [inline config/deploy.yml.sample]: Though we don't want to use YML for this after all?

**Comment by sirupsen** [comment]
@glebm are you running the new version of docker? IIRC it falls back to that if AUFS is not available.

File: .ruby-gemset (hunk 1 above)
Comment (haileys) [inline .ruby-gemset]: For developing the app locally you should run `script/bootstrap` instead of `bundle`. This will make sure to properly isolate the environment for the app. Sorry this wasn't made more clear in the README.

File: config/deploy.yml.sample (hunk 6 above)
Comment (sirupsen) [inline config/deploy.yml.sample]: Yeah you are right @glebm, I'd like to have this in bash for now.

File: script/generate-ssl-certificate [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+cp $DOMAIN.key $DOMAIN.key.org
+fail_if_error $?
+
+# Strip the password so we don't have to type it every time we restart Apache
+openssl rsa -in $DOMAIN.key.org -out $DOMAIN.key -passin env:PASSPHRASE
+fail_if_error $?
+
+# Generate the cert (good for 10 years)
+openssl x509 -req -days 3650 -in $DOMAIN.csr -signkey $DOMAIN.key -out $DOMAIN.crt
+fail_if_error $?
```
Comment (bouk) [inline script/generate-ssl-certificate]: I don't think this file should be in the repo, you can just look up how to generate a cert online

File: script/provision-basic (hunk 5 above)
Comment (glebm) [inline script/provision-basic]: Is this package for AUFS to work with docker?

File: script/generate-ssl-certificate (hunk 7 above)
Comment (sirupsen) [inline script/generate-ssl-certificate]: Perhaps we should default to http for ease of deployment?

File: script/provision-basic (hunk 5 above)
Comment (sirupsen) [inline script/provision-basic]: Shouldn't be. I'm not exactly sure where AUFS was installed on our deployment. I'll look into this tomorrow, I'm heading out soon

Comment (glebm) [inline script/provision-basic]: I am not quite sure what to do with this one, as there is no default unversioned package, and uname is not reliable

File: .ruby-gemset (hunk 1 above)
Comment (sirupsen) [inline .ruby-gemset]: The reason it's not clear in the README is because it's run as part of the provisioning in Vagrant—you shouldn't really have to care.

File: script/provision-basic (hunk 5 above)
Comment (bouk) [inline script/provision-basic]: linux-image-extra is for AUFS yes

Bouke

File: .ruby-gemset (hunk 1 above)
Comment (haileys) [inline .ruby-gemset]: @Sirupsen well you have to after someone's changed the Gemfile or Gemfile.lock

File: script/generate-ssl-certificate (hunk 7 above)
Comment (glebm) [inline script/generate-ssl-certificate]: Generating certs is not straightforward. This makes a basic install way easier.

Comment (haileys) [inline script/generate-ssl-certificate]: Generating an SSL certificate isn't really something that should be automated in the provision script IMO.

Besides, self-signed certificates don't provide any considerable security over plaintext HTTP so I don't see a lot of value in this unless it's signed by a recognised CA.

**Comment by glebm** [comment]
I've updated the PR and added a section on deployment to Readme.

File: README.md [hunk 8]
```diff
@@ -73,5 +73,36 @@ forwarded from the Vagrant-managed VM.
 
 ## Deployment
 
-Coming soon. For now `script/provision-production` should do the trick on an
-Ubuntu Raring x64 box.
+Deploy to an an Ubuntu Raring x64 box.
```
Comment (sirupsen) [inline README.md]: Are you running Raring? Let's recommend Saucy. Once 14.04 LTS comes out with a kernel that supports Docker that's what we should recommend long term.

File: README.md [hunk 9]
```diff
@@ -73,5 +73,36 @@ forwarded from the Vagrant-managed VM.
...
+
+First, copy your public SSH key to your home directory on the target server:
```
Comment (sirupsen) [inline README.md]: Imo., this is not really in the scope of deploying Contestrus.

File: config/deploy.sh.sample [hunk 10]
```diff
@@ -0,0 +1,2 @@
+DEPLOY_USER=app
```
Comment (sirupsen) [inline config/deploy.sh.sample:1]: You have both `.sample` and without the sample suffix. I think we should only have the samples checked in, and add those without the `.sample` suffix to gitignore. This adds an extra step to deploy, but people should check this out anyhow.

File: script/provision-basic [hunk 11]
```diff
@@ -34,8 +34,15 @@ apt-get install -y \
...
+  wget \
+  bsdtar \
+  curl \
   zlib1g-dev \
-  linux-image-extra-`uname -r`
+  aufs-tools \
+  lxc \
+  linux-virtual
+
+apt-get install linux-image-extra-`uname -r` || apt-get install linux-image-extra-3.11.0-13-generic
```
Comment (sirupsen) [inline script/provision-basic:45]: :clap:

File: config/deploy.sh.sample (hunk 10 above)
Comment (glebm) [inline config/deploy.sh.sample:1]: This was erroneously checked, thanks! No extra step is necessary because deploy-env does provide defaults.

**Comment by glebm** [comment]
`script/provision` now works, Readme updated

File: README.md (hunk 8 above)
Comment (glebm) [inline README.md]: Running 13.10, left the name over from before (can never remember those right), updated.

File: README.md (hunk 9 above)
Comment (glebm) [inline README.md]: Updated all of that now. Turns out there was a scripts/provision already, just had to get it working.

**Comment by sirupsen** [comment]
Thanks @glebm! Merging.
